openhands/integrations/github/github_service.py

```diff
@@ -615,6 +617,82 @@ async def get_branches(self, repository: str) -> list[Branch]:
 
         return all_branches
 
+    async def get_paginated_branches(
+        self, repository: str, page: int = 1, per_page: int = 30
+    ) -> PaginatedBranchesResponse:
+        """Get branches for a repository with pagination"""
+        url = f'{self.BASE_URL}/repos/{repository}/branches'
+
+        params = {'per_page': str(per_page), 'page': str(page)}
+        response, headers = await self._make_request(url, params)
+
+        branches: list[Branch] = []
+        for branch_data in response:
+            # Extract the last commit date if available
+            last_push_date = None
+            if branch_data.get('commit') and branch_data['commit'].get('commit'):
+                commit_info = branch_data['commit']['commit']
+                if commit_info.get('committer') and commit_info['committer'].get(
+                    'date'
+                ):
+                    last_push_date = commit_info['committer']['date']
+
+            branch = Branch(
+                name=branch_data.get('name'),
+                commit_sha=branch_data.get('commit', {}).get('sha', ''),
+                protected=branch_data.get('protected', False),
+                last_push_date=last_push_date,
+            )
+            branches.append(branch)
+
+        # Parse Link header to determine if there's a next page
+        has_next_page = False
+        if 'Link' in headers:
+            link_header = headers['Link']
+            has_next_page = 'rel="next"' in link_header
+
+        return PaginatedBranchesResponse(
+            branches=branches,
+            has_next_page=has_next_page,
+            current_page=page,
+            per_page=per_page,
+            total_count=None,  # GitHub doesn't provide total count in branch API
+        )
+
+    async def search_branches(
```

Comment:
@openhands the search branches implementation for github provider is incorrect; we should be passing the user's partial query as a parameter to the github api when searching for the branch

github may not support a dedicated endpoint for this as per `https://docs.github.com/en/rest/branches/branches?apiVersion=2022-11-28`

1. maybe check if there is a graphql endpoint instead
2. write the graphql query inside `openhands/integrations/github/queries.py`
3. rewrite the search endpoint with the graphql query (check the existing pattern for calling graphql queries)
4. write a test script (but do not commit this script), where you use the $GITHUB_TOKEN in your env to call the graphql query to see that its working as expected